components/proxy/src/main/kotlin/com/hotels/styx/routing/db/StyxObjectStore.kt

```diff
@@ -15,28 +15,100 @@
+import java.util.concurrent.Executors
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.Consumer
 
 /**
  * Styx Route Database.
  */
-class StyxObjectStore<T> : ObjectStore<T> {
-    private val objects = ConcurrentHashMap<String, Record<T>>();
+class StyxObjectStore<T> : ObjectStoreSnapshot<T> {
+    private val objects: AtomicReference<PMap<String, Record<T>>> = AtomicReference(HashTreePMap.empty())
+    private val watchers = CopyOnWriteArrayList<ChangeWatcher<T>>()
+    private val executor = Executors.newSingleThreadExecutor()
 
-    override fun get(name: String?): Optional<T> {
-        return Optional.ofNullable(objects.get(name))
+    override fun get(name: String): Optional<T> {
+        return Optional.ofNullable(objects().get(name))
                 .map { it.payload }
     }
 
+    fun insert(key: String, payload: T) {
+        insert(key, setOf(), payload)
+    }
+
+    fun remove(key: T) {
+        queue {
+            val nextVersion = objects().minus(key)
+            objects.set(nextVersion)
+            notifyWatchers(nextVersion)
+        }
+    }
+
+    /**
+     * Returns a Publisher that emits an event at any modification.
+     *
+     * Watch activates on subscription only.
+     * Watch removed on unsubscription.
+     */
+    fun watch(): Publisher<ObjectStoreSnapshot<T>> {
+        return Flux.push { sink ->
+            val watcher = object : ChangeWatcher<T> {
+                override fun accept(objectStoreSnapshot: ObjectStoreSnapshot<T>) {
+                    sink.next(objectStoreSnapshot)
+                }
+            }
+
+            sink.onDispose {
+                watchers.remove(watcher)
+            }
+
+            watchers.add(watcher)
+
+            queue {
+                sink.next(ObjectStoreSnapshot { key ->
+                    Optional
+                            .ofNullable(objects().get(key))
+                            .map { it.payload }
+                })
+            }
+        }
+    }
+
+    internal fun watchers() = watchers.size
+
+    private fun objects() = objects.get()
+
+    private fun queue(task: () -> Unit) {
+        executor.submit(task)
+    }
+
     private fun insert(key: String, tags: Set<String>, payload: T) {
-        objects.put(key, Record(key, tags, payload))
+        queue {
+            val nextVersion = objects().plus(key, Record(key, tags, payload))
+            objects.set(nextVersion)
+            notifyWatchers(nextVersion)
+        }
     }
 
-    override fun insert(key: String, record: T) {
-        insert(key, setOf(), record)
+    private fun notifyWatchers(objectsV2: PMap<String, Record<T>>) {
+        watchers.forEach { listener ->
+            listener.accept(snapshot(objectsV2))
+        }
     }
 
-    data class Record<T>(val key: String, val tags:Set<String>, val payload: T)
+    private fun snapshot(snapshot: PMap<String, Record<T>>) = object : ObjectStoreSnapshot<T> {
```

Comment:
I believe Kotlin supports SAM conversions only for Java interoperability. Not for within Kotlin:

> Also note that this feature works only for Java interop; since Kotlin has proper function types, automatic conversion of functions into implementations of Kotlin interfaces is unnecessary and therefore unsupported.

(from: https://kotlinlang.org/docs/reference/java-interop.html#sam-conversions)

This suggests you cannot substitute `ObjectStore` return type for Lambda. Or do you see something else?

Reply:
Okay, it now looks like this:

    private fun snapshot(snapshot: PMap<String, Record<T>>) = ObjectStore<T> { key: String ->
        Optional
                .ofNullable(snapshot.get(key))
                .map { it.payload }
    }